### ISSUE TITLE
Unit tests for gokit examples

### ISSUE THREAD

**Issue by danielvinson** [go-kit/kit#1175]
### What would you like?

I have been working recently on building gokit microservices and found it difficult to get off the ground in writing unit tests for services.  I think it would be extremely helpful to add unit tests to at least one example or add a new example with tests.

I'd be happy to open a PR for this, but I'd love some guidance on where people would see value here.  I think `stringsvc3` would be the best place to start here, but I'd also be happy to make a brand new one.

**Comment by peterbourgon** [comment]
My instinct is that this is out-of-scope for Go kit itself. There just isn't a single, well-defined way to do them.

**Comment by sagikazarmark** [comment]
I think I agree with Peter on this one, here is why:

If you want to write unit tests for the business logic: well, it's not that complicated and I'm not sure what it would demonstrate exactly as the business logic for a service is entirely Go kit independent.

Writing tests for the endpoint (+service) layer is possible, but probably not that important as you would test that integration with transport level tests as well.

Transport level tests are rather e2e/integration tests for me (depending on how they are constructed). They are usually messy and I don't really see them demonstrating any Go kit specific scenarios either.

In short, I don't really see what these tests would demonstrate apart from testing the dummy business logic and some integration.

There is one reason why having some tests would be beneficial though: to ensure that the examples are functional.

**Comment by danielvinson** [comment]
> My instinct is that this is out-of-scope for Go kit itself. There just isn't a single, well-defined way to do them.

I understand that offering an example might lead people in one direction when there are many others to choose from.  In that case I'll write some tests as a separate repo.  Would you be interested in linking back to that in this repo?

**Comment by peterbourgon** [comment]
> many companies (including mine) like having 100% test coverage . . .

Testing is a risk mitigating strategy, and has an asymptotic cost/benefit curve. Very few organizations exist at a point on the spectrum where 100% coverage delivers net benefit, and those organizations that do should certainly not be using Go 😉 I encourage you to reconsider a 100% test coverage target, in my experience it's actively harmful.

**Comment by peterbourgon** [comment]
> it is extremely hard to test gokit middleware in an isolated way.

Your tests shouldn't worry about the code in Go kit itself, we test that ourselves! 😉 But if you have your own e.g. Endpoint middleware, it should be straightforward to test in a totally isolated way. What problems are you having?

**Comment by danielvinson** [comment]
> Your tests shouldn't worry about the code in Go kit itself, we test that ourselves! 😉 But if you have your own e.g. Endpoint middleware, it should be straightforward to test in a totally isolated way. What problems are you having?

I appreciate the advice.  My problem isn't that this is impossible, it was just a few days of work to find a way to test this properly.  I just want to save the next few people who attempt this a bit of time :)
 
I've got a middleware which handles auth with an external API which requires a token refresh every hour, then sets that as a header using a ClientBefore function.  While this ends up as fairly simple code, it is very hard to test that the middleware attempts to refresh the token when it expires.  Where I ended up is that I built a mock for the external API using `httptest.Server`, then I use a mock NoOp endpoint which I wrap in the middleware.  The only gap in coverage is for the `ClientBefore` function which sets the header, which I believe I don't need to worry about.

**Comment by sagikazarmark** [comment]
I'm not sure I understand 100% what this middleware does, but in my experience testing these things in isolation is somewhat easier: you don't need an actual API call for the test. Mocking out the API is one option, hiding the whole thing behind an interface, providing an alternative implementation is another.

Some might say it's extreme, but I would even hide time behind an interface, so that your middleware doesn't have any side effects and can be tested in complete isolation. You can use [clockwork](https://github.com/jonboulle/clockwork) to do that.

This of course has the downside of not testing the actual integration (ie. the API returns what it ought to). While most people argue this shouldn't be the client's job, it's good to keep in mind that not having automated tests for these kinds of things requires you to manually test them.

**Comment by danielvinson** [comment]
Thank you for the advice guys, I really appreciate it.
